### PR TITLE
Add SOEM 1.4.0

### PR DESCRIPTION
**Pull request by seanyen**
Checklist

- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example)
- [x] Source is from official source
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged)
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details)
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there

### CODE REVIEW

**Comment by seanyen** [comment]
cc @conda-forge/staged-recipes this is ready for review and merge.

File: recipes/soem/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  host:
+    - pthread-stubs                      # [not win]
+  run:
+    - pthread-stubs                      # [not win]
+
+test:
+  commands:
+    - test -f ${PREFIX}/include/soem/ethercat.h  # [not win]
+    - test -f ${PREFIX}/bin/slaveinfo  # [not win]
+    - test -f ${PREFIX}/lib/libsoem.a  # [not win]
```
Comment (isuruf) [inline recipes/soem/meta.yaml]: Can you build shared libraries?

Comment (seanyen) [reply recipes/soem/meta.yaml]: I don't know how far the mileage would be, but let me try. :)

**Comment by seanyen** [comment]
@isuruf I patched the upstream and convert it to shared library build.

**Comment by seanyen** [comment]
cc @conda-forge/staged-recipes this is ready for review and merge. thanks!

File: recipes/soem/build-shlib.patch [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
++
++#if defined _WIN32 || defined __CYGWIN__
++  #ifdef __GNUC__
++    #define ETHERCATMAIN_EXPORT __attribute__ ((dllexport))
++    #define ETHERCATMAIN_IMPORT __attribute__ ((dllimport))
++  #else
++    #define ETHERCATMAIN_EXPORT __declspec(dllexport)
++    #define ETHERCATMAIN_IMPORT __declspec(dllimport)
++  #endif
++  #ifdef ETHERCATMAIN_BUILDING_LIBRARY
```
Comment (isuruf) [inline recipes/soem/build-shlib.patch]: Is this defined?

**Pull request merged**